pkg/controllers/workapplier/controller.go

```diff
@@ -0,0 +1,1126 @@
+	// by the bug would have an index with invalid identifier strings; Fleet is aware of this
+	// situation and it is still safe to handle manifests with such an erred index as in later
+	// steps the code will attempt to rebuild the association between the manifest object and
+	// the object from the member cluster.
+
+	// Prepare the bundles.
+	bundles := prepareManifestProcessingBundles(work)
+
+	// Pre-process the manifests to apply.
+	//
+	// In this step, Fleet will:
+	// a) decode the manifests; and
+	// b) write ahead the manifest processing attempts; and
+	// c) remove any applied manifests left over from previous runs.
+	if err := r.preProcessManifests(ctx, bundles, work, expectedAppliedWorkOwnerRef); err != nil {
+		klog.ErrorS(err, "Failed to pre-process the manifests", "work", workRef)
+		return ctrl.Result{}, err
+	}
+
+	// Process the manifests.
+	r.processManifests(ctx, bundles, work, expectedAppliedWorkOwnerRef)
+	// Track the availability information.
+	r.trackInMemberClusterObjAvailability(ctx, bundles, &workRef)
+
+	trackWorkApplyLatencyMetric(work)
+
+	// Refresh the status of the Work object.
+	if err := r.refreshWorkStatus(ctx, work, bundles); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Refresh the status of the AppliedWork object.
+	if err := r.refreshAppliedWorkStatus(ctx, appliedWork, bundles); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// If the Work object is not yet available, reconcile again in 5 seconds.
+	if !isWorkObjectAvailable(work) {
+		return ctrl.Result{RequeueAfter: availabilityCheckRequeueAfter}, nil
+	}
+	// Otherwise, reconcile again in 3 minutes for drift detection purposes.
+	return ctrl.Result{RequeueAfter: driftCheckRequeueAfter}, nil
+}
+
+// preProcessManifests pre-processes manifests for the later ops.
+func (r *Reconciler) preProcessManifests(
+	ctx context.Context,
+	bundles []*manifestProcessingBundle,
+	work *fleetv1beta1.Work,
+	expectedAppliedWorkOwnerRef *metav1.OwnerReference,
+) error {
+	// Decode the manifests.
+	// Run the decoding in parallel to boost performance.
+	//
+	// This is concurrency safe as the bundles slice has been pre-allocated.
+
+	// Prepare a child context.
+	// Cancel the child context anyway to avoid leaks.
+	childCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	doWork := func(pieces int) {
+		bundle := bundles[pieces]
+		if bundle.applyErr != nil {
+			// Skip a manifest if it cannot be processed.
+			return
+		}
+
+		gvr, manifestObj, err := r.decodeManifest(bundle.manifest)
+		// Build the identifier. Note that this would return an identifier even if the decoding
+		// fails.
+		bundle.id = buildWorkResourceIdentifier(pieces, gvr, manifestObj)
+		if err != nil {
+			klog.ErrorS(err, "Failed to decode the manifest", "ordinal", pieces, "work", klog.KObj(work))
+			bundle.applyErr = fmt.Errorf("failed to decode manifest: %w", err)
+			bundle.applyResTyp = ManifestProcessingApplyResultTypeDecodingErred
+			return
```

Comment:
This won't return any error back to the caller, thus there are bundles that have no gvr/ManifestObj. However,  those are used extensively as pointers in the rest part of the controller logic. It seems that this can lead to Nullptr panic?

Reply:
I see that there are checks like "bundle.applyErr != nil" on some places but I am not sure if it covers all cases.  Maybe we can add a step after preProcessManifests to remove those from the bundle?

Reply:
Hi Ryan! Yeah, at this moment the applier would skip a bundle for the processing step if it has failed the pre-processing; for the bundle removal part, a complication is that since we need to report occurrences of such manifests (cannot be decoded/malformed) to the users, if we remove them from the bundles before processing, in the status refreshing part we would need to take extra steps to make sure that they are re-incorporated, which can also be error-prone I fear.

Would you prefer if I:

a) shuffle the list of the bundles after pre-processing to make sure that bundles that failed pre-processing are set in the back, and before processing slice the array to leave them out; or
b) check for nil GVR/manifest object right before processing and throw an unexpected error?